External WAL directory and datapage checksums for PostgreSQL

moench-tegeder opened the pull request:
Since almost forever, PostgreSQL allows storing it's transaction log
(WAL, write ahead log) outside of the data directory (this improves
performance on certain setups, or may just be required by local
policy. The newly added "waldir" option on states.postgres_initdb.present()
and modules.postgresql.initdb() is passed to initdb (the default is
still to keep the transaction log inside the data directory).
As part of the big renaming in PostgreSQL 10 the old "xlog directory"
has been renamed to "wal directory", this patch uses the new name only.

Sind PostgreSQL 9.3 (which is the oldest supported release as of
time of writing), there is support for checksumming the database
on the block level. This can only be enabled when creating the
cluster (this is, at initdb time), so it should be available in
modules and states. This patch adds it in both places.

Documentation for both mini-features has been added to the doc strings.

### What does this PR do?

### What issues does this PR fix or reference?

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

Yes/No

### Commits signed with GPG?

Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

moench-tegeder committed:
External WAL directory and datapage checksums for PostgreSQL

Since almost forever, PostgreSQL allows storing it's transaction log
(WAL, write ahead log) outside of the data directory (this improves
performance on certain setups, or may just be required by local
policy. The newly added "waldir" option on states.postgres_initdb.present()
and modules.postgresql.initdb() is passed to initdb (the default is
still to keep the transaction log inside the data directory).
As part of the big renaming in PostgreSQL 10 the old "xlog directory"
has been renamed to "wal directory", this patch uses the new name only.

Sind PostgreSQL 9.3 (which is the oldest supported release as of
time of writing), there is support for checksumming the database
on the block level. This can only be enabled when creating the
cluster (this is, at initdb time), so it should be available in
modules and states. This patch adds it in both places.

Documentation for both mini-features has been added to the doc strings.

moench-tegeder commented:
The versionadded tag has been set to XXX, as it's up to your release planning if and when you merge this.

Commit:
Merge branch 'develop' into postgres_options

gtmanfred commented on salt/modules/postgres.py:
```diff
@@ -3109,8 +3122,21 @@ def datadir_init(name,
+    .. note:
+
+        The ``waldir`` and ``checksum`` options have been added in
+        version XXX. The ``checksum`` option requires PostgreSQL 9.3
```
This will be added in `Fluorine`, you can use the `.. versionadded: Fluorine` macro for rst.

moench-tegeder committed:
move feature history into versionadded tag

moench-tegeder committed:
move feature history in execution module, too

while here, provide helpful note about data page checksums being
supported since PostgreSQL 9.3.

moench-tegeder commented:
Ah right. I shouldn't fix documentation while overtired. Gave it a little improvement, too. Thanks for noticing.

gtmanfred commented on salt/modules/postgres.py:
```diff
@@ -3109,8 +3122,22 @@ def datadir_init(name,
+
+    checksums
+        If True, the cluster will be created with data page checksums.
+
```
Should be Fluorine, like the element in the periodic table

gtmanfred reviewed (changes requested):
Just need to fix the spelling and this should be good

moench-tegeder committed:
fix spelling of version name "Fluorine"

moench-tegeder commented:
Drats. One of the few words spelled with "uo".

gtmanfred commented on salt/states/postgres_initdb.py:
```diff
@@ -61,6 +63,19 @@ def present(name,
+        The transaction log (WAL) directory (default is to keep WAL
+        inside the data directory)
+
+        .. versionadded:: Flourine
```
One last thing :P these guys need to be fixed too.

moench-tegeder committed:
really fix all my Fluorine typos

rallytime commented on salt/modules/postgres.py:
```diff
@@ -199,6 +199,8 @@ def _run_initdb(name,
         password=None,
         encoding='UTF8',
         locale=None,
+        waldir=None,
```
Can you move these new kwargs to the end of the list? We don't want to cause an API change in case users are passing positional args.

rallytime commented on salt/modules/postgres.py:
```diff
@@ -3079,6 +3090,8 @@ def datadir_init(name,
         encoding='UTF8',
         locale=None,
+        waldir=None,
+        checksums=False,
```
Same comment here.

rallytime reviewed (changes requested):
A couple of small requests from me.

moench-tegeder committed:
put new kwargs at the end

avoiding breakage for people working against common wisdom by using
positional args in these places

moench-tegeder commented:
sure, while I'm here.
And while grep failed me yesterday, today it found even more misspellings: https://github.com/saltstack/salt/pull/48052 :)

moench-tegeder commented:
Hi again,
is there anything I can do here? Unless I'm missing something, thw kwargs are now in order?

rallytime commented:
Thanks for the bump. It looks great now. Thank you!
